src/librustc_typeck/check/demand.rs

```diff
@@ -47,8 +53,48 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 assert!(obligations.is_empty());
             },
             Err(e) => {
-                self.report_mismatched_types(origin, expected, actual, e);
+                self.report_mismatched_types(origin, expected, actual, e).emit();
+            }
+        }
+    }
+
+    fn check_ref(&self, expr: &hir::Expr, checked_ty: Ty<'tcx>,
+                 expected: Ty<'tcx>) -> Option<String> {
+        match (&checked_ty.sty, &expected.sty) {
+            (&ty::TyRef(_, x_mutability), &ty::TyRef(_, y_mutability)) => {
+                // check if there is a mutability difference
```

Comment:
Hum... The only case scenario I imagined was very similar:

```rust
let x = 0u32;
bar(&x);
fn bar(x: &mut i32) {}
```

But I didn't think of such a case. I wonder if it's possible to know if a variable can become mutable in the current context or not...